Timer rewrite phase 3: drivers

andyross opened the pull request:
Just like the last (still active) PR #10303 this builds on top of the previous one and adds extra patches.

In this case, it takes advantage of the much simpler timer driver API to actually implement much simpler drivers for ARM SysTick and Intel HPET.  Seriously, these are like 4x smaller now.

They support all the features the old ones did, and now work with tickless pervasively (and enable tickless by default, as there's essentially no downside anymore).

Please review.  Also please test if you can, especially on SysTick hardware (I've only run tests on Qemu and the two boards I have on my desk).

I have similar rewrites in progress for Xtensa and RISC-V MTIME (which are very similar in structure to HPET) and will append those here as I finish them.

While this can be merged as a single unit, it's probably best to merge the older pull request first to smooth the transition.  I'll rebase this one as soon as the older one goes in.

Commit:
drivers/timer: New ARM SysTick driver

Newer, and much smaller driver written to the new timer API.  Supports
all the features the old one did (including shutting off the clock
when clock_always_on is disabled), should be faster in practice, and
should be significantly more accurate due to the "lost cycle" trick
applied in z_clock_set_timeout().

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/arm/qemu_cortex_m3: Make it tickless by default

The SysTick driver is working with tickless now.  Make this our first
guinea pig.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/benchmarks: Disable TICKLESS_KERNEL for "no tick" tests

These tests was written to try to eliminate timer interrupts, but the
mechanism chosen (setting TICKS_PER_SECOND to 1) is dangerously
susceptible to overflow conditions on systems with fast cycle counters
and high timeout durations.

Actually the tests pass just fine if you use a conventional tick rate
and use a tickless-capable driver (which eliminates interrupts too,
which is the whole point), but there's no easy way in kconfig to do an
"if" to select that condition for capable systems only.  Just disable
tickless to keep the same behavior.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer: Add a standard workaround for known qemu issues

Qemu doesn't like tickless.  By default[1] it tries to be realtime as
vied by the host CPU -- presenting read values from hardware cycle
counters and interrupt timings at the appropriate real world clock
times according to whatever the simulated counter frequency is.  But
when the host system is loaded, there is always the problem that the
qemu process might not see physical CPU time for large chunks of time
(i.e. a host OS scheduling quantum -- generally about the same size as
guest ticks!) leading to lost cycles.

When those timer interrupts are delivered by the emulated hardware at
fixed frequencies without software intervention, that's not so bad:
the work the guest has to do after the interrupt generally happens
synchronously (because the qemu process has just started running) and
nothing notices the dropout.

But with tickless, the interrupts need to be explicitly programmed by
guest software!  That means the driver needs to be sure it's going to
get some real CPU time within some small fraction of a Zephyr tick of
the right time, otherwise the computations get wonky.

The end result is that qemu tends to work with tickless well on an
unloaded/idle run, but not in situations (like sanitycheck) where it
needs to content with other processes for host CPU.

So, add a flag that drivers can use to "fake" tickless behavior when
run under qemu (only), and enable it (only!) for the small handful of
tests that are having trouble.

[1] There is an -icount feature to implement proper cycle counting at
the expense of real-world-time correspondence.  Maybe someday we might
get it to work for us.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer: Add more default stubs

Many drivers won't need to implement z_clock_idle_exit() or
sys_clock_disable(). Make those weak stubs too.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer: Select tickless via driver kconfig flag

Add a TICKLESS_CAPABLE kconfig variable which is used by the kernel to
select tickless mode's default automatically on drivers that support
it (rather than having to set the default per-board).  Select it from
the ARM SysTick and Intel HPET drivers.

Also remove the old qemu_cortex_m3 default settings which this
replaces.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/context: Limit no-tick-during-irq-load to !TICKLESS

When TICKLESS_KERNEL is enabled, the current time in ticks is based on
a hardware counter and not interrupt delivery (which is the whole
point of tickless), so irq-locking does not prevent time from
advancing.  Disable this test in that configuration.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/interrupt: Shrink very long k_busy_wait() argument

This test was written with an outrageously long timeout of 25 seconds.
That blows right through the 32 bit cycle counter on qemu_cortex_m3[1]
and produces an essentially random delay instead of the desired
number, causing a hang with the new SysTick driver in tickless mode.

Push the number down so it doesn't overflow.  The root cause, though,
is that k_busy_wait() can take arguments it can't handle.  It ought to
have an outer loop or something so that it can spin for INT_MAX
milliseconds correctly.

[1] Which has a 12MHz clock rate.  Many hardware implementations are
much faster still.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer/hpet: Completely new, simplified, tickless-capable driver

Rewritten along the lines of ARM SysTick.  Implements only the new,
simplified API.  MUCH smaller.  Works with tickless pervasively.  No
loss of functionality.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer: Reworked NRF driver with tickless support

Reworked using the older hardware interface code, but with an
implementation of the new API only.  Much smaller & simpler.

As yet, tested (manually) only on a nrf52_pca10056 board.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix race in k_sched_time_slice_set()

If this function is itself interrupted by a timeslice event, the
slicing state can be corrupted.  Just re-use the scheduler lock
instead of using a new spinlock; this is a low-latency function that
won't deadlock.  Found by inspection.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Update with some fixes to HPET...

...and a reworked nrf_rtc_timer driver that really wants review from Nordic folks.  It works for me on a nrf52_pca10056 at least, but needs validation on broader hardware.  The size should make review easy at least.

andyross commented:
Rebased.  Fixed an issue on nrf where the driver and kernel don't care about the symbol name of the ISR, but one test does.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -523,87 +86,73 @@ int z_clock_driver_init(struct device *device)
 
-	irq_disable(NRF5_IRQ_RTC1_IRQn);
+	/* Round up to next tick boundary */
+	cyc = ticks * CYC_PER_TICK + counter_sub(t, last_count);
```
What about potential overflow in these calculations?

andyross replied:
Ticks gets clamped to [0,MAX_TICKS] in the code above this spot, which is calibrated to prevent that.

andyross replied:
To be clear: that's about preventing overflow beyond 24 bits of cycles in a single timeout.

There's no need to prevent overflow in the calculation of the next timeout value itself, the cycle counter is cyclic and just rolls over.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -523,87 +86,73 @@ int z_clock_driver_init(struct device *device)
+		cyc += CYC_PER_TICK;
+	}
 
+	set_comparator(cyc);
```
I am not sure if while tick to cycle inculcation will work correctly if we set tick equal to hw clock cycle.
It looks that announce will be always scheduled 1 clock cycle earlier than expected in such case.

andyross replied:
Not sure I understand the point.  The cyc value here is the time of the last expiration (last_count) plus some integer number of CYC_PER_TICK increments.  Unless the timer skips a cycle when the comparator fires, the ticks should be perfectly in-phase and evenly scheduled (modulo interrupt delivery slop, of course).

andyross replied:
Oh wait, I see what you're saying: in tickless, you HAVE to use the current time, because counting interrupts isn't sufficient.  But the math is guaranteed not to underrun.

pizi-nordic reviewed (commented):
I am still afraid of using `now()` time in `z_clock_set_timeout()`.
Even with the lock there is small probability of slipping over clock cycle during each clock programming operation. Could we use last announcement as reference here? In such case we will have to read the `now()` only when new timeout will be added to timeout queue, and all further operations will be error-free.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -1,513 +1,76 @@
-	/* Calculate time to program into RTC */
-	rtc_away = (RTC_MASK - RTC_COUNTER);
-	rtc_away = rtc_away > RTC_HALF ? RTC_HALF : rtc_away;
+#define MIN_DELAY 32
```
Do we really need this minimal delay?
In current implementation, this basically reduces timer resolution, which is a thing we would like to avoid.

andyross replied:
You have to have a minimum value greater than zero here, otherwise it's possible that the computation of the "next" value based on the "current" timer counter will compute a value that is in the "past" by the timer it arrives in the comparator register.

I picked 1/10th of a tick to be conservative, but given the frequencies here something much lower is probably fine.  You guys are the experts :)

andyross commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -523,87 +86,73 @@ int z_clock_driver_init(struct device *device)
-	unsigned int key;
+#ifdef CONFIG_TICKLESS_KERNEL
+	ticks = ticks == K_FOREVER ? MAX_TICKS : ticks;
+	ticks = max(min(ticks - 1, (s32_t)MAX_TICKS), 0);
```
Because our API is werid.  Historically we've passed a "1" to set_timeout() to indicate "at the next tick" (and then had to play games with a _TICK_ALIGN symbol needlessly).  But in the code here we want to compute a delta from current time rounded up to the next tick boundary, so we start from ticks-1.

andyross commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -1,513 +1,76 @@
-{
-	u32_t rtc_away, sys_away = 0;
+#define COUNTER_MAX 0x00ffffff
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC	\
```
I'm trying to deprecate that as an API, actually.  It's dangerous due to precision loss and has a weirdness (it has to look like a function call) just to deal with HPET (which has to query the frequency at runtime).  The only code that uses it is in the drivers, and I'd prefer they do the math themselves.

Commit:
drivers/timer: New, tickless-capable RISC-V machine timer driver

Rewritten driver along the lines of all the other new drivers,
implementing the new timer API.  Structurally, the machine timer is an
up-counter with comparator, so it works broadly the same way HPET and
NRF do.  The quirk here is that it's a 64 bit counter, which needs a
little more care.

Unlike the other timer reworks, this driver has grown by a few lines
as it used to be very simple.  But in exchange, we get full tickless
support on the platform.

Fixes #10609 in the process (the 64 bit timer registers are unlatched
for sub-word transfers, so you have to use careful ordering).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

nategraff-sifive commented:
@andyross I ran the dining philosophers sample on HiFive 1 in both tickfull and tickless configurations and it appears to be running happily with the new driver.

Commit:
drivers/timer: New xtensa timer with tickless support

Rewritten Xtensa CCOUNT driver along the lines of all the other new
drivers.  The new API permits much smaller code.

Notably: The Xtensa counter is a 32 bit up-counter with a comparator
register.  It's in some sense the archetype of this kind of timer as
it's the simplest of the bunch (everything else has quirks: NRF is
very slow and 24 bit, HPET has a runtime frequency detection, RISC-V
is 64 bit...).  I should have written this one first.

Note also that this includes a blacklist of the xtensa architecture on
the tests/driver/ipm test.  I'm getting spurious failures there where
a k_sem_take() call with a non-zero timeout is being made out of the
console output code in interrupt context.  This seems to have nothing
to do with the timer; I suspect it's because the old timer drivers
would (incorrectly!) call z_clock_announce() in non-interrupt context
in some contexts (e.g. "expiring really soon").  Apparently this test
(or something in the IPM or Xtensa console code) was somehow relying
on that on Xtensa.  But IPM is a Quark thing and there's no particular
reason to run this test there.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebased.  Add an xtensa driver (tested so far only on qemu, my esp32 is acting funny at the moment).  Also adds a workaround for nrf52_bsim along the lines of the existing qemu one (i.e. disable tickless when the we're in a host emulation environment that uses host time as the reference).

At this point I think this whole path is complete.  I don't plan on doing rewrites of the Pulpino driver (for which I don't have hardware) or ARC (for which my only hardware is... cumbersome).  Nios II would be possible as it has a qemu variant, but looking at it the hardware seems to be a regular interrupt kind of thing and wouldn't support tickless anyway (rather: it's possible the hardware does, but the existing driver doesn't use those bits).

aescolar commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -38,5 +38,11 @@ config LOG_PROCESS_THREAD
 
 endif # LOG
 
+# Like other emulation environments that rely on host timekeeping (and
```
Nono, the nrf52_bsim does not use the host time for anything. So the host scheduling should not affect in any way Zephyr's execution. What problem did you see? (I can help you debug it)

andyross replied:
Hm... how does it synthesize cycle counts out of the timer device then?  For example: if you spin waiting on the next tick, how does it know that time is passing?

Regardless, if you remove that patch you'll see failures out of sanitycheck when tickless is enabled.

andyross replied:
Oooh.... having typed that and walked away, I'm like 80% sure that's exactly the situation.  In "ticked" mode, the kernel responds with what amounts to an interrupt count when you ask it for the current time.  In tickless, it checks the current hardware counter for the delta from the last known tick instead.

If in native_posix, the "infinite clock rate" trick is going to mess that up.   In which case the workaround remains the same but I need to update the explanation to be correct :)  Let me dig a bit.

aescolar replied:
> Presumably the clock looks frozen 

Exactly, the CPU/SW execution runs in 0 time (inside one simulation "delta cycle"), and simulated time only continues when `k_cpu_idle()` or `k_busy_wait()` are called.

aescolar replied:
This fixes tests/kernel/tickless/tickless_concept
```diff
diff --git a/tests/kernel/tickless/tickless_concept/src/main.c b/tests/kernel/tickless/tickless_concept/src/main.c
index d0b39b9..636013e 100644
--- a/tests/kernel/tickless/tickless_concept/src/main.c
+++ b/tests/kernel/tickless/tickless_concept/src/main.c
@@ -33,7 +33,7 @@ static struct k_thread tdata[NUM_THREAD];
        do {                                   \
                u32_t t = k_uptime_get_32();   \
                while (t == k_uptime_get_32()) \
-                       posix_halt_cpu();      \
+                       k_busy_wait(50);      \
        } while (0)
 #else
 #define ALIGN_MS_BOUNDARY()                   \
```
Explanation:
The ALIGN_MS_BOUNDARY hack for the POSIX arch relied on having the ticker interrupt (between others) awaking the CPU.
But if you halt the CPU with the tick disabled, nothing would awake it for an arbitrary length of time. In that case, just replacing the CPU half with a small delay in each iteration of the loop is a good solution (50µs in this case, but any small, arbitrary number would do)

aescolar replied:
Same fix for tests/kernel/sleep/
```diff
diff --git a/tests/kernel/sleep/src/main.c b/tests/kernel/sleep/src/main.c
index 8bebca0..19814af 100644
--- a/tests/kernel/sleep/src/main.c
+++ b/tests/kernel/sleep/src/main.c
@@ -73,7 +73,7 @@ static void align_to_tick_boundary(void)
        while (k_uptime_get_32() == tick) {
                /* Busy wait to align to tick boundary */
 #if defined(CONFIG_ARCH_POSIX)
-               posix_halt_cpu();
+               k_busy_wait(50);
 #endif
        }
```

aescolar replied:
And more of the same for tests/kernel/sched/schedule_api
```diff
diff --git a/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c b/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c
index 6e51321..4668cf6 100644
--- a/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c
+++ b/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c
@@ -104,7 +104,7 @@ void test_slice_reset(void)
                t32 = k_uptime_get_32();
                while (k_uptime_get_32() - t32 < HALF_SLICE_SIZE) {
 #if defined(CONFIG_ARCH_POSIX)
-                       posix_halt_cpu(); /*sleep until next irq*/
+                       k_busy_wait(50); /*sleep until next irq*/
 #else
                        ;
 #endif
```

nategraff-sifive commented on drivers/timer/riscv_machine_timer.c:
```diff
@@ -1,117 +1,135 @@
-	z_clock_announce(1);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	u32_t now = mtime();
+	u32_t dticks = ((u32_t)now - (u32_t)last_count) / CYC_PER_TICK;
```
I think these u32_t's are causing overflow bugs when I run some tests on hardware for an extended period of time. I haven't sat down and really thought through the math, but I get weird timer behavior coinciding with mtime values larger than 32 bits. Changing these variables to u64_t's fixes it.

andyross replied:
Huh.  It certainly looks right.  How long an extended period are we talking?  Can you rerun the test with both variants and an _ASSERT to check that they are equivalent?  Does it work if you do the "(now-last_count)/CYC_PER_TICK" in 64 bit but then cast for dticks?

Compiler bugs happen too -- do you have a more recent toolchain to try than the SDK?

nategraff-sifive replied:
On the platform I'm testing on, the machine timer clock is driven at 12.5 MHz, and the misbehavior happens consistently after just over 5 minutes and 30 seconds, the same time that `mtime` rolls over to 33 bits.

I'll try the tests you suggest and see what I can figure out.

nategraff-sifive replied:
Setting an __ASSERT does detect a discrepancy between the argument to `set_mtimecmp()` when the mtime rollover happens. The deciding factor looks like it might be when truncating `now` to 32-bits.

The code that works is
```
u64_t now = mtime();
u32_t dticks = (u32_t)((now - last_count) / CYC_PER_TICK);
```

I was building on SDK 0.9.3. Updating to 0.9.4-rc3 still exhibits the problem, so at least it's not  a compiler bug solved between those two releases.

andyross commented:
Rebased.  Fixed a nasty initialization bug on ARM that was exposed by the mps2 target.  Added @nategraff-sifive's suggested fix for the 64 bit rollover condition (that I still don't really see, looking at it -- though this version is still optimizable by the compiler to the 32 bit one we want, so there's not reason not to do it this way).

Should be good to merge at this point.  Please review.

tyrantlizard77 reviewed (commented):
Curious if we're going to refactor the LOAPIC timer driver at some point.

andyross commented on boards/arm/mps2_an385/Kconfig.board:
```diff
@@ -4,6 +4,11 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+# Note: the MPS2 is a real hardware board, but this target runs under
```
I kind of agree.  I mean, if qemu was perfect I'd say feeding it a build for whatever hardware it's emulating would be fine, but given that we have workarounds in place for virtualization, it makes things needlessly complicated.  That said, the actual MPS2 board doesn't seem like a very common piece of hardware -- maybe we should rename this qemu_mps2 or something and deprecate support for the real board?

nashif replied:
will create an alias, the hw itself exists https://os.mbed.com/platforms/ARM-MPS2.

andyross commented on tests/kernel/common/prj.conf:
```diff
@@ -2,3 +2,4 @@ CONFIG_ZTEST=y
 CONFIG_PRINTK=y
 CONFIG_LOG=y
 CONFIG_POLL=y
+CONFIG_QEMU_TICKLESS_WORKAROUND=y
```
This is the kconfig that the test sets.  Basically "I need the qemu workaround".  Most tests without host-load-dependent timing behavior work fine with the unaltered tickless driver.

andyross commented:
APIC timer refactoring is coming as part of x86_64, FWIW.  In fact the need for that is sort of the initial germ that got me started on this whole timer saga.

andyross commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -523,87 +86,73 @@ int z_clock_driver_init(struct device *device)
 
 	/* Clear the event flag and possible pending interrupt */
-	RTC_CC_EVENT = 0;
+	nrf_rtc_event_clear(RTC, 0);
```
I copied this from the earlier driver honestly.  But will fix.

andyross commented:
Rebased and pushed a final version.  All code is present, all tests are passing.  All needed workaround (e.g. for nrf54_bsim's spin loops) are already in the tree.  I think all criticism is addressed.  Needs final review.

carlescufi reviewed (approved):
Approving in lieu of @pizi-nordic who is away this week

nategraff-sifive reviewed (approved):
RISC-V LGTM
